### PR TITLE
Editor

### PR DESCRIPTION
**Pull request by Sirmorrison**
### Resolves

Issue #85

### What's new
What's new
* This PR Adds a decred widget test App to this project.
* This Pr added a custom Editor.
* Custom Editor has a paste and clear button
* Custom Editor has option of hiding the paste and clear button
* Custom Editor has is required option.
* Adds color animation when active.
* Custom Editor has line and was tested with VerifyMessage dialogue
* Custom Editor is design similar to https://materializecss.com/text-inputs.html
 
_A detailed description of what this pull request adds_

### Relevant screenshots or logs
ezgif com-video-to-gif(4)

### CODE REVIEW

File: ui/decredmaterial/editor.go [hunk 1]
```diff
@@ -26,6 +30,24 @@ type Editor struct {
...
+	editorMaterial Editor
+	flexWidth      float32
+	editor         *widget.Editor
+
+	IsVisibleBtn     bool
+	pasteBtnMaterial IconButton
+	pasteBtnWidget   *widget.Button
+
+	clearBtMaterial IconButton
+	clearBtnWidget  *widget.Button
```
Comment (oluwandabira) [inline ui/decredmaterial/editor.go]: Remove `Btn` from the names.
`IsVisibleBtn` says nothing about what this boolean control. What "is visible"?, he editor, the custom buttons?

File: ui/decredmaterial/editor.go [hunk 2]
```diff
@@ -26,6 +30,24 @@ type Editor struct {
 	shaper text.Shaper
 }
 
+type EditorCustom struct {
+	theme *Theme
```
Comment (oluwandabira) [inline ui/decredmaterial/editor.go]: Don't save the theme, create the widgets the need and save them instead

File: ui/decredmaterial/editor.go [hunk 3]
```diff
@@ -26,6 +30,24 @@ type Editor struct {
...
+	//title is the title of the editor input field
+	title     string
```
Comment (oluwandabira) [inline ui/decredmaterial/editor.go]: Why even have a comment if the field is not exported? Is there a reason it is not exported?

File: ui/decredmaterial/editor.go [hunk 4]
```diff
@@ -62,3 +114,95 @@ func (e Editor) Layout(gtx *layout.Context, editor *widget.Editor) {
...
+		e.editor.SetText(data)
+	}
+	for e.clearBtnWidget.Clicked(gtx) {
+		e.editor.SetText("")
+	}
+
+	for _, evt := range e.editor.Events(gtx) {
+		switch evt.(type) {
+		case widget.ChangeEvent:
+			e.editorMaterial.HintColor = e.theme.Color.Hint
```
Comment (oluwandabira) [inline ui/decredmaterial/editor.go]: Why is this necessary?

**Review by oluwandabira — CHANGES_REQUESTED** [review]

File: ui/decredmaterial/editor.go (hunk 3 above)
Comment (Sirmorrison) [reply ui/decredmaterial/editor.go]: Yes. Because the title does uses the hint text as its value. It does not need its value changed or written to externally. This is in a bid to replicate what we have on https://materializecss.com/text-inputs.html

File: ui/decredmaterial/editor.go (hunk 4 above)
Comment (Sirmorrison) [reply ui/decredmaterial/editor.go]: Since we have codes that had to use the events on the editor on handle_input.go, i felt its easier we eliminate those repeating codes by adding it here instead.

Comment (oluwandabira) [reply ui/decredmaterial/editor.go]: But the editor's `HintColor` can't change since its not exported.

File: ui/decredmaterial/editor.go (hunk 3 above)
Comment (oluwandabira) [reply ui/decredmaterial/editor.go]: Over there the hint (placeholder) can be different from the title.

File: ui/decredmaterial/editor.go [hunk 5]
```diff
@@ -26,6 +30,31 @@ type Editor struct {
...
+	// horizontal.
+	SingleLine bool
+
+	pasteBtnMaterial IconButton
+	pasteBtnWidget   *widget.Button
+
+	clearBtMaterial IconButton
+	clearBtnWidget  *widget.Button
+}
+
```
Comment (oshorefueled) [inline ui/decredmaterial/editor.go]: Is there a reason we have an Editor and a Custom Editor? Why not have just one editor?

File: ui/dialogs.go [hunk 6]
```diff
@@ -252,114 +251,29 @@ func (win *Window) verifyMessageDiag() {
 					win.vFlexSB(
 						rigid(func() {
 							inset := layout.Inset{
-								Top: unit.Dp(10),
+								Top:    unit.Dp(10),
+								Bottom: unit.Dp(10),
```
Comment (oshorefueled) [inline ui/dialogs.go]: Why the modification here? not sure what the effect of the change is.

File: ui/page.go [hunk 7]
```diff
@@ -68,7 +68,6 @@ func (win *Window) addPages() {
 	win.pages[PageRestore] = win.RestorePage
 	win.pages[PageSend] = win.SendPage
 	win.pages[PageSignMessage] = win.SignMessagePage
-
 }
```
Comment (oshorefueled) [inline ui/page.go:87]: you should avoid making changes to code that's barely related to what you're working on.

File: ui/widgets.go [hunk 8]
```diff
@@ -38,7 +38,6 @@ type inputs struct {
 		text   string
 		button widget.Button
 	}
-
```
Comment (oshorefueled) [inline ui/widgets.go]: Avoid making changes to code that's barely related to what you're working on.

**Review by oshorefueled — CHANGES_REQUESTED** [review]

File: ui/decredmaterial/editor.go [hunk 9]
```diff
@@ -51,14 +177,74 @@ func (e Editor) Layout(gtx *layout.Context, editor *widget.Editor) {
...
 	} else {
 		macro.Add()
 	}
 	paint.ColorOp{Color: e.Color}.Add(gtx.Ops)
-	editor.PaintCaret(gtx)
+	e.EditorWdgt.PaintCaret(gtx)
 	stack.Pop()
 }
+
+func (e Editor) EditorWdgtLine(gtx *layout.Context) {
```
Comment (oshorefueled) [inline ui/decredmaterial/editor.go]: There's no need to abbreviate Widget to "wdgt" here. The length of the method name is just fine. Removing an i and e doesn't seem to do too much shortening.

File: ui/decredmaterial/icon.go [hunk 10]
```diff
@@ -7,7 +7,7 @@ import (
 	"image"
 	"image/color"
 	"image/draw"
-	_ "image/png"
+	_ "image/png" // fixes unknown format error when trying to decode png images
```
Comment (oshorefueled) [inline ui/decredmaterial/icon.go]: this comment should be rephrased. It makes png a decodable format, it doesn't fix an error.

**Review by oshorefueled — CHANGES_REQUESTED** [review]

**Review by oshorefueled — APPROVED** [review]

**Review by oluwandabira — APPROVED** [review]
